File: dotnet/SeleniumCore/ChromeDriver.cs
using System.Collections.Generic;

namespace OpenQA.Selenium;

/// <summary>
/// A Chrome session that resolves, downloads-or-caches, and launches its OWN
/// chromedriver via the shared engine — no driver on PATH, no Grid. Mirrors
/// Selenium 4.x's <c>OpenQA.Selenium.Chrome.ChromeDriver</c>: the no-arg
/// constructor "just works", and the options constructor threads
/// <c>goog:chromeOptions</c>-style capabilities. The launched driver process is
/// stopped on <see cref="RemoteWebDriver.Quit"/> / <see cref="RemoteWebDriver.Dispose"/>.
/// </summary>
public class ChromeDriver : RemoteWebDriver
{
    /// <summary>Launch chromedriver and start a default Chrome session.</summary>
    public ChromeDriver() : this((IDictionary<string, object?>?)null)
    {
    }

    /// <summary>
    /// Launch chromedriver and start a Chrome session with the given options
    /// (e.g. <c>new Dictionary&lt;string, object?&gt; { ["goog:chromeOptions"] =
    /// new Dictionary&lt;string, object?&gt; { ["args"] = new List&lt;object?&gt;
    /// { "--headless=new" } } }</c>). <c>browserName=chrome</c> is set automatically.
    /// </summary>
    public ChromeDriver(IDictionary<string, object?>? options)
        : this(EnsureChromeDriver(), options)
    {
    }

    /// <summary>
    /// Launch chromedriver and start a Chrome session from a mainstream
    /// <see cref="OpenQA.Selenium.Chrome.ChromeOptions"/> (or any
    /// <see cref="DriverOptions"/>): its <see cref="DriverOptions.ToCapabilities"/>
    /// supplies the capabilities. This is the Selenium 4.x <c>new ChromeDriver(options)</c>
    /// entry point.
    /// </summary>
    public ChromeDriver(DriverOptions options)
        : this(EnsureChromeDriver(), (options ?? throw new System.ArgumentNullException(nameof(options))).ToDictionary())
    {
    }

    private ChromeDriver(DriverProcess proc, IDictionary<string, object?>? options)
        : base(proc.Url, WithChrome(options), null, false)
    {
        AdoptDriver(proc);
    }

    private static DriverProcess EnsureChromeDriver() =>
        EnsureDriver("chrome") ?? throw new WebDriverException("could not resolve/launch chromedriver", -1);

    private static IDictionary<string, object?> WithChrome(IDictionary<string, object?>? options)
    {
        var caps = new Dictionary<string, object?> { ["browserName"] = "chrome" };
        if (options != null)
        {
            foreach (var kv in options)
            {
                caps[kv.Key] = kv.Value;
            }
        }
        return caps;
    }
}

File: java/src/main/java/org/openqa/selenium/ChromeDriver.java
package org.openqa.selenium;

import java.util.Map;

/**
 * A Chrome session that resolves, downloads-or-caches, and launches its OWN
 * chromedriver via the shared engine — no driver on PATH, no Grid. Mirrors
 * Selenium 4.x's {@code org.openqa.selenium.chrome.ChromeDriver}: the no-arg
 * constructor "just works", and the options constructor threads
 * {@code goog:chromeOptions}-style capabilities. The launched driver process is
 * stopped on {@link #quit()}.
 */
public class ChromeDriver extends RemoteWebDriver {

    /** Launch chromedriver and start a default Chrome session. */
    public ChromeDriver() {
        this(Map.of());
    }

    /**
     * Launch chromedriver and start a Chrome session with the given options
     * (e.g. {@code Map.of("goog:chromeOptions", Map.of("args", List.of("--headless=new")))}).
     * {@code browserName=chrome} is set automatically.
     */
    public ChromeDriver(Map<String, Object> options) {
        super(ensureChromeDriver(), withChrome(options), null, false);
    }

    /**
     * Launch chromedriver and start a Chrome session configured by an upstream
     * {@link org.openqa.selenium.chrome.ChromeOptions} (or any
     * {@link org.openqa.selenium.Capabilities}). Mirrors Selenium 4.x's
     * {@code new ChromeDriver(chromeOptions)}.
     */
    public ChromeDriver(org.openqa.selenium.Capabilities options) {
        super(ensureChromeDriver(), withChrome(new java.util.HashMap<>(options.asMap())), null, false);
    }

    private static DriverProcess ensureChromeDriver() {
        DriverProcess proc = ensureDriver("chrome");
        if (proc == null) {
            throw new WebDriverException("could not resolve/launch chromedriver", -1);
        }
        return proc;
    }

    private static Map<String, Object> withChrome(Map<String, Object> options) {
        java.util.Map<String, Object> caps = new java.util.HashMap<>();
        caps.put("browserName", "chrome");
        if (options != null) {
            caps.putAll(options);
        }
        return caps;
    }
}
